src/cdef_apply.rs

```diff
@@ -291,7 +238,9 @@ pub(crate) unsafe fn rav1d_cdef_brow<BD: BitDepth>(
                     | (*noskip_row.offset(0))[0] as c_uint;
                 y_lvl = frame_hdr.cdef.y_strength[cdef_idx as usize];
                 uv_lvl = frame_hdr.cdef.uv_strength[cdef_idx as usize];
-                flag = ((y_lvl != 0) as c_int + (((uv_lvl != 0) as c_int) << 1)) as Backup2x8Flags;
+                flag = Backup2x8Flags::from_bits_truncate(
+                    (y_lvl != 0) as u32 + (((uv_lvl != 0) as u32) << 1),
+                );
```

Comment:
It is ugly.  It generates slightly different code vs. `if`s + `.insert`s, though the difference is unlikely to matter at all.  Ideally I think this would use `bool` bitfields, but Rust doesn't support that well.  Maybe just leave it as is to be safe?  Also, `bitflags!` has added `::from_bits_retain` now, which you can use here instead of `::from_bits_truncate`.

Reply:
I found what I think is a more elegant and perfectly optimal solution:

```rust
impl Backup2x8Flags {
    pub const fn union_all<const N: usize>(a: [Self; N]) -> Self {
        let mut this = Self::empty();
        let mut i = 0;
        while i < N {
            this = this.union(a[i]);
            i += 1;
        }
        this
    }

    pub const fn select(&self, select: bool) -> Self {
        Self::from_bits_retain(self.bits() * select as u8)
    }

    pub const fn from_bitfields(y: bool, uv: bool) -> Self {
        Self::union_all([Self::Y.select(y), Self::UV.select(uv)])
    }
}
```

it [compiles down to pretty optimal assembly](https://godbolt.org/z/67s9P87q9):

```asm
Backup2x8Flags::from_bitfields:
        lea     eax, [rsi + rsi]
        or      al, dil
        ret
```

Reply:
```rust
    pub const fn select(&self, select: bool) -> Self {
        if select {
            *self
        } else {
            Self::empty()
        }
    }
```

This also still gets optimal codegen, but is easier to read.

Reply:
Actually, the `union_all` way doesn't optimize well sometimes, so this is the best I think:

```rust
impl Backup2x8Flags {
    pub const fn select(&self, select: bool) -> Self {
        if select {
            *self
        } else {
            Self::empty()
        }
    }
}
```

```suggestion
                flag = Backup2x8Flags::Y.select(y_lvl != 0) | Backup2x8Flags::UV.select(uv_lvl != 0)
```

Reply:
nice, that looks a lot better!